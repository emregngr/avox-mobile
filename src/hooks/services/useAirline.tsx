import { useQuery } from '@tanstack/react-query'
import { useEffect, useMemo, useState } from 'react'

import { getAllAirlines } from '@/services/airlineService'
import useLocaleStore from '@/store/locale'
import type { Airline } from '@/types/feature/airline'
import { useDebounce } from '@/utils/common/useDebounce'
import { parseFilterRange } from '@/utils/feature/parseFilterRange'

const ITEMS_PER_PAGE = 20

export function useAirline() {
  const { selectedLocale } = useLocaleStore()
  const [searchTerm, setSearchTerm] = useState<string>('')
  const [filters, setFilters] = useState<any>({})
  const [page, setPage] = useState<number>(1)
  const [isSearchLoading, setIsSearchLoading] = useState<boolean>(false)
  const debouncedSearch = useDebounce<string>(searchTerm, 500)

  const {
    data: allAirlinesData,
    error,
    isLoading,
  } = useQuery({
    queryFn: () => getAllAirlines(selectedLocale),
    queryKey: ['allAirlines', selectedLocale],
    staleTime: 5 * 60 * 1000,
  })

  const applyFilters = (airlines: Airline[]): Airline[] => {
    if (!airlines) return []

    return airlines.filter(airline => {
      if (!airline) return false

      if (filters?.region) {
        const airlineRegion = airline?.operations?.region?.toLowerCase()
        const filterRegion = filters?.region?.toLowerCase()

        if (airlineRegion !== filterRegion) {
          return false
        }
      }

      if (filters?.businessType) {
        const businessType = airline?.operations?.businessType?.toLowerCase()
        const filterBusinessType = filters?.businessType?.toLowerCase()

        if (businessType !== filterBusinessType) {
          return false
        }
      }

      if (filters?.businessModel) {
        const businessModel = airline?.operations?.businessModel?.toLowerCase()
        const filterBusinessModel = filters?.businessModel?.toLowerCase()

        if (businessModel !== filterBusinessModel) {
          return false
        }
      }

      if (filters?.alliance) {
        const alliance = airline?.operations?.alliance?.toLowerCase()
        const filterAlliance = filters?.alliance?.toLowerCase()

        if (alliance !== filterAlliance) {
          return false
        }
      }

      if (filters?.minSkytraxRating && airline?.operations?.skytraxRating) {
        if (airline?.operations?.skytraxRating < filters?.minSkytraxRating) {
          return false
        }
      }

      const airlineRangeChecks: {
        filterKey: keyof typeof filters
        isYear?: boolean
        jsonPath: (a: Airline) => number | undefined | null
      }[] = [
        {
          filterKey: 'foundingYear',
          isYear: true,
          jsonPath: a =>
            a.companyInfo?.foundingYear
              ? parseInt(String(a?.companyInfo?.foundingYear), 10)
              : undefined,
        },
        { filterKey: 'passengerCapacity', jsonPath: a => a?.companyInfo?.passengerCapacity },
        { filterKey: 'employeeCount', jsonPath: a => a?.companyInfo?.employeeCount },
        { filterKey: 'destinationCount', jsonPath: a => a?.network?.destinationCount },
        { filterKey: 'destinationCountries', jsonPath: a => a?.network?.destinationCountries },
        { filterKey: 'domesticConnections', jsonPath: a => a?.network?.domesticConnections },
        {
          filterKey: 'internationalConnections',
          jsonPath: a => a?.network?.internationalConnections,
        },
        { filterKey: 'totalAirplane', jsonPath: a => a?.fleet?.totalAirplane },
        { filterKey: 'averageAgeYears', jsonPath: a => a?.fleet?.averageAgeYears },
        { filterKey: 'airplaneTypeCount', jsonPath: a => a?.fleet?.airplaneTypeCount },
      ]

      for (const check of airlineRangeChecks) {
        const filterValueStr = filters?.[check.filterKey] as string
        if (filterValueStr) {
          const parsed = parseFilterRange(filterValueStr)
          if (parsed) {
            let itemValue = check?.jsonPath(airline)
            if (itemValue === undefined || itemValue === null || isNaN(Number(itemValue))) {
              return false
            }
            itemValue = Number(itemValue)
            if (itemValue < parsed.min || itemValue > parsed.max) {
              return false
            }
          }
        }
      }
      return true
    })
  }

  const filteredAirlines = useMemo(() => {
    if (!allAirlinesData) return []

    let filtered = allAirlinesData

    if (debouncedSearch.trim()) {
      const searchLower = debouncedSearch.toLowerCase().trim()

      const exactMatches = filtered.filter(
        airline =>
          airline.iataCode?.toLowerCase() === searchLower ||
          airline.icaoCode?.toLowerCase() === searchLower,
      )

      if (exactMatches.length > 0) {
        filtered = exactMatches
      } else {
        filtered = filtered.filter(airline => {
          const name = airline?.name?.toLowerCase() || ''
          const city = airline?.operations?.hub?.city?.toLowerCase() || ''
          const country = airline?.operations?.country?.toLowerCase() || ''
          const iata = airline?.iataCode?.toLowerCase() || ''
          const icao = airline?.icaoCode?.toLowerCase() || ''

          return (
            name.includes(searchLower) ||
            city.includes(searchLower) ||
            country.includes(searchLower) ||
            iata.includes(searchLower) ||
            icao.includes(searchLower)
          )
        })
      }
    }

    return applyFilters(filtered)
  }, [allAirlinesData, debouncedSearch, filters])

  const paginatedAirlines = useMemo(() => {
    if (!filteredAirlines) return []
    return filteredAirlines.slice(0, page * ITEMS_PER_PAGE)
  }, [filteredAirlines, page])

  const hasMore = paginatedAirlines.length < filteredAirlines.length

  const loadMore = () => {
    if (hasMore && !isLoading) {
      setPage(prev => prev + 1)
    }
  }

  useEffect(() => {
    if (searchTerm !== debouncedSearch) {
      setIsSearchLoading(true)
    } else {
      setTimeout(() => setIsSearchLoading(false), 300)
    }
  }, [searchTerm, debouncedSearch])

  useEffect(() => {
    setPage(1)
  }, [debouncedSearch, filters])

  return {
    error,
    filteredCount: filteredAirlines?.length,
    filters,
    hasMore,
    isLoading,
    isSearchLoading,
    loadMore,
    paginatedAirlines,
    searchTerm,
    setFilters,
    setSearchTerm,
  }
}
