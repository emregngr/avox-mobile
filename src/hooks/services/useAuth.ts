import { getApp } from '@react-native-firebase/app'
import type { FirebaseAuthTypes } from '@react-native-firebase/auth'
import { getAuth, getIdToken, onAuthStateChanged } from '@react-native-firebase/auth'
import type { QueryClient } from '@tanstack/react-query'
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { router } from 'expo-router'
import { Alert } from 'react-native'
import Toast from 'react-native-toast-message'

import { getLocale } from '@/locales/i18next'
import {
  handleLogout,
  sendPasswordResetLink,
  signInWithApple,
  signInWithEmail,
  signInWithGoogle,
  signUpWithEmail,
} from '@/services/authService'
import { login, logout, register, social } from '@/store/auth'
import useThemeStore from '@/store/theme'
import type { AuthCredentials, LoginCredentials, RegisterCredentials } from '@/types/feature/auth'
import { Logger } from '@/utils/common/logger'

const app = getApp()
const auth = getAuth(app)

const handleAuthSuccess = async <T extends AuthCredentials>(
  userCredential: FirebaseAuthTypes.UserCredential,
  authFunction: (data: T & { token: string }) => Promise<void>,
  authData: T,
  queryClient: QueryClient,
) => {
  try {
    const token = await getIdToken(userCredential.user, true)
    await authFunction({ ...authData, token })
    queryClient.invalidateQueries({ queryKey: ['user'] })
  } catch (error) {
    Logger.breadcrumb('Failed to handle auth success', 'error', error as Error)
  }
}

export const useUser = () =>
  useQuery({
    queryFn: () =>
      new Promise(resolve => {
        const unsubscribe = onAuthStateChanged(auth, userState => {
          unsubscribe()
          resolve(userState)
        })
      }),
    queryKey: ['user'],
    staleTime: Infinity,
  })

export const useEmailRegister = () => {
  const queryClient = useQueryClient()
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: async ({ email, firstName, lastName, password }: RegisterCredentials) =>
      await signUpWithEmail({ email, firstName, lastName, password }),
    onError: () =>
      Alert.alert(
        getLocale('error'),
        getLocale('emailRegisterFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      ),
    onSuccess: async (userCredential, variables) => {
      await handleAuthSuccess(
        userCredential,
        register,
        {
          email: variables.email,
          firstName: variables.firstName,
          lastName: variables.lastName,
          password: variables.password,
          token: '',
        },
        queryClient,
      )
    },
  })
}

export const useEmailLogin = () => {
  const queryClient = useQueryClient()
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: async ({ email, password }: LoginCredentials) =>
      await signInWithEmail({ email, password }),
    onError: () =>
      Alert.alert(
        getLocale('error'),
        getLocale('emailLoginFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      ),
    onSuccess: async (userCredential, variables) => {
      await handleAuthSuccess(
        userCredential,
        login,
        {
          email: variables.email,
          password: variables.password,
        },
        queryClient,
      )
    },
  })
}

const useSocialLogin = (
  signInMethod: () => Promise<FirebaseAuthTypes.UserCredential>,
  provider: 'google' | 'apple',
) => {
  const queryClient = useQueryClient()
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: async () => await signInMethod(),
    onError: () =>
      Alert.alert(
        getLocale('error'),
        getLocale('socialLoginFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      ),
    onSuccess: async userCredential => {
      await handleAuthSuccess(userCredential, social, { provider }, queryClient)
    },
  })
}

export const useGoogleLogin = () => useSocialLogin(signInWithGoogle, 'google')

export const useAppleLogin = () => useSocialLogin(signInWithApple, 'apple')

export const useLogout = () => {
  const queryClient = useQueryClient()
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: async () => await handleLogout(),
    onError: () => {
      Alert.alert(
        getLocale('error'),
        getLocale('logoutFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      )
    },
    onSuccess: async () => {
      await logout()
      queryClient.setQueryData(['user'], null)
      queryClient.removeQueries()
    },
  })
}

export const useForgotPassword = () => {
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: sendPasswordResetLink,
    onError: () =>
      Alert.alert(
        getLocale('error'),
        getLocale('passwordResetLinkSentMessageFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      ),
    onSuccess: () => {
      Toast.show({
        text1: getLocale('successful'),
        text2: getLocale('passwordResetLinkSentMessage'),
        type: 'success',
      })
      router.back()
    },
  })
}
