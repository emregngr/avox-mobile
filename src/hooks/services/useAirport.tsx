import { useQuery } from '@tanstack/react-query'
import { useEffect, useMemo, useState } from 'react'

import { getAllAirports } from '@/services/airportService'
import useLocaleStore from '@/store/locale'
import type { Airport } from '@/types/feature/airport'
import { useDebounce } from '@/utils/common/useDebounce'
import { parseFilterRange } from '@/utils/feature/parseFilterRange'

const ITEMS_PER_PAGE = 20

export function useAirport() {
  const { selectedLocale } = useLocaleStore()
  const [searchTerm, setSearchTerm] = useState<string>('')
  const [filters, setFilters] = useState<any>({})
  const [page, setPage] = useState<number>(1)
  const [isSearchLoading, setIsSearchLoading] = useState<boolean>(false)
  const debouncedSearch = useDebounce<string>(searchTerm, 500)

  const {
    data: allAirportsData,
    error,
    isLoading,
  } = useQuery({
    queryFn: () => getAllAirports(selectedLocale),
    queryKey: ['allAirports', selectedLocale],
    staleTime: 5 * 60 * 1000,
  })

  const applyFilters = (airports: Airport[]): Airport[] => {
    if (!airports) return []

    return airports.filter(airport => {
      if (!airport) return false

      if (filters?.region) {
        const airportRegion = airport?.operations?.region?.toLowerCase()
        const filterRegion = filters?.region?.toLowerCase()

        if (airportRegion !== filterRegion) {
          return false
        }
      }

      if (filters?.airportType) {
        const airportType = airport?.operations?.airportType?.toLowerCase()
        const filterAirportType = filters?.airportType?.toLowerCase()

        if (airportType !== filterAirportType) {
          return false
        }
      }

      const airportServices = airport?.facilities?.services?.map(s => s?.toLowerCase()) || []

      const serviceKeywords: Record<string, string[]> = {
        hasCarRental: ['car rental'],
        hasChildrensPlayArea: ["children's play area"],
        hasDutyFree: ['duty-free shops'],
        hasHealthServices: ['health services'],
        hasHotels: ['hotels'],
        hasLounges: ['lounges'],
        hasPrayerRoom: ['prayer rooms'],
        hasRestaurants: ['restaurants'],
      }

      for (const serviceName in serviceKeywords) {
        if (filters?.[serviceName] === true) {
          const keywords = serviceKeywords?.[serviceName]

          const hasMatchingService = airportServices?.some(serviceString =>
            keywords?.some((keyword: string) => serviceString?.includes?.(keyword)),
          )

          if (!hasMatchingService) {
            return false
          }
        }
      }

      if (filters?.is24Hour === true && !airport?.operations?.is24Hour) return false
      if (filters?.is24Hour === false && airport?.operations?.is24Hour) return false

      if (filters?.hasMetro === true && !airport?.facilities?.hasMetro) return false
      if (filters?.hasMetro === false && airport?.facilities?.hasMetro) return false

      if (filters?.freeWifi === true && !airport?.facilities?.freeWifi) return false
      if (filters?.freeWifi === false && airport?.facilities?.freeWifi) return false

      if (filters?.minGoogleRating && airport?.facilities?.googleMapsRating) {
        if (airport?.facilities?.googleMapsRating < filters?.minGoogleRating) {
          return false
        }
      }

      const rangeChecks: {
        filterKey: keyof typeof filters
        isYear?: boolean
        jsonPath: (a: Airport) => number | undefined | null
      }[] = [
        {
          filterKey: 'foundingYear',
          isYear: true,
          jsonPath: a =>
            a?.airportInfo?.foundingYear
              ? parseInt(String(a?.airportInfo?.foundingYear), 10)
              : undefined,
        },
        { filterKey: 'passengerCapacity', jsonPath: a => a?.infrastructure?.passengerCapacity },
        { filterKey: 'employeeCount', jsonPath: a => a?.airportInfo?.employeeCount },
        { filterKey: 'destinationCount', jsonPath: a => a?.flightOperations?.destinationCount },
        {
          filterKey: 'destinationCountries',
          jsonPath: a => a.flightOperations?.destinationCountries,
        },
        {
          filterKey: 'domesticConnections',
          jsonPath: a => a?.flightOperations?.domesticConnections,
        },
        {
          filterKey: 'internationalConnections',
          jsonPath: a => a?.flightOperations?.internationalConnections,
        },
        { filterKey: 'elevationFt', jsonPath: a => a?.operations?.location?.elevationFt },
        {
          filterKey: 'passengerCapacity',
          jsonPath: a => a?.infrastructure?.passengerCapacity,
        },
        { filterKey: 'baggageCapacity', jsonPath: a => a?.infrastructure?.baggageCapacity },
        { filterKey: 'terminalCount', jsonPath: a => a?.infrastructure?.terminalCount },
        {
          filterKey: 'terminalAreaHectares',
          jsonPath: a => a?.infrastructure?.terminalAreaHectares,
        },
        { filterKey: 'airportAreaHectares', jsonPath: a => a?.infrastructure?.airportAreaHectares },
        { filterKey: 'apronCount', jsonPath: a => a?.infrastructure?.apronCount },
        { filterKey: 'runwayCount', jsonPath: a => a?.infrastructure?.runwayCount },
        { filterKey: 'mainRunwayLengthM', jsonPath: a => a?.infrastructure?.runways?.lengthM },
        { filterKey: 'towerHeightM', jsonPath: a => a?.infrastructure?.towerHeightM },
        {
          filterKey: 'parkingCapacityVehicles',
          jsonPath: a => a?.facilities?.parkingCapacityVehicles,
        },
        { filterKey: 'loungeCount', jsonPath: a => a?.facilities?.loungeCount },
        { filterKey: 'securityQueueTime', jsonPath: a => a?.facilities?.securityQueueTime },
        { filterKey: 'checkinTimeAvg', jsonPath: a => a?.facilities?.checkinTimeAvg },
      ]

      for (const check of rangeChecks) {
        const filterValueStr = filters?.[check.filterKey] as string
        if (filterValueStr) {
          const parsed = parseFilterRange(filterValueStr)
          if (parsed) {
            let itemValue = check?.jsonPath(airport)
            if (itemValue === undefined || itemValue === null || isNaN(Number(itemValue))) {
              return false
            }
            itemValue = Number(itemValue)
            if (itemValue < parsed.min || itemValue > parsed.max) {
              return false
            }
          }
        }
      }
      return true
    })
  }

  const filteredAirports = useMemo(() => {
    if (!allAirportsData) return []
    let filtered = allAirportsData

    if (debouncedSearch.trim()) {
      const searchLower = debouncedSearch.toLowerCase().trim()

      const exactMatches = filtered.filter(
        airport =>
          airport.iataCode?.toLowerCase() === searchLower ||
          airport.icaoCode?.toLowerCase() === searchLower,
      )

      if (exactMatches.length > 0) {
        filtered = exactMatches
      } else {
        filtered = filtered.filter(airport => {
          const name = airport?.name?.toLowerCase() || ''
          const city = airport?.operations?.location?.city?.toLowerCase() || ''
          const country = airport?.operations?.country?.toLowerCase() || ''
          const iata = airport?.iataCode?.toLowerCase() || ''
          const ident = airport?.icaoCode?.toLowerCase() || ''

          return (
            name.includes(searchLower) ||
            city.includes(searchLower) ||
            country.includes(searchLower) ||
            iata.includes(searchLower) ||
            ident.includes(searchLower)
          )
        })
      }
    }

    return applyFilters(filtered)
  }, [allAirportsData, debouncedSearch, filters])

  const paginatedAirports = useMemo(() => {
    if (!filteredAirports) return []
    return filteredAirports.slice(0, page * ITEMS_PER_PAGE)
  }, [filteredAirports, page])

  const hasMore = paginatedAirports.length < filteredAirports.length

  const loadMore = () => {
    if (hasMore && !isLoading) {
      setPage(prev => prev + 1)
    }
  }

  useEffect(() => {
    if (searchTerm !== debouncedSearch) {
      setIsSearchLoading(true)
    } else {
      setTimeout(() => setIsSearchLoading(false), 300)
    }
  }, [searchTerm, debouncedSearch])

  useEffect(() => {
    setPage(1)
  }, [debouncedSearch, filters])

  return {
    error,
    filteredCount: filteredAirports?.length,
    filters,
    hasMore,
    isLoading,
    isSearchLoading,
    loadMore,
    paginatedAirports,
    searchTerm,
    setFilters,
    setSearchTerm,
  }
}
