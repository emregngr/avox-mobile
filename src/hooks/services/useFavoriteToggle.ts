import { router } from 'expo-router'
import { useCallback, useMemo } from 'react'

import { useAddFavorite, useIsFavorite, useRemoveFavorite } from '@/hooks/services/useFavorite'
import useAuthStore from '@/store/auth'
import type { FavoriteItemType } from '@/types/feature/favorite'
import { Logger } from '@/utils/common/logger'

export function useFavoriteToggle({ id, type }: FavoriteItemType) {
  const { isAuthenticated } = useAuthStore()

  const { isPending: isAddFavoritePending, mutateAsync: addFavoriteMutation } = useAddFavorite()
  const { isPending: isRemoveFavoritePending, mutateAsync: removeFavoriteMutation } =
    useRemoveFavorite()

  const isFavorite = useIsFavorite({ id, type })

  const isPending = useMemo(
    () => isAddFavoritePending || isRemoveFavoritePending,
    [isAddFavoritePending, isRemoveFavoritePending],
  )

  const handleFavoritePress = useCallback(async (): Promise<void> => {
    if (!isAuthenticated) {
      router.replace('/auth')
      return
    }

    if (isPending) return

    try {
      if (isFavorite) {
        await removeFavoriteMutation({ id, type })
      } else {
        await addFavoriteMutation({ id, type })
      }
    } catch (error) {
      Logger.breadcrumb('Failed to toggle favorite', 'error', error as Error)
    }
  }, [
    isAuthenticated,
    isPending,
    isFavorite,
    id,
    type,
    addFavoriteMutation,
    removeFavoriteMutation,
  ])

  return { handleFavoritePress, isFavorite, isPending }
}
