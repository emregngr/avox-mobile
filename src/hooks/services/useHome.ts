import { useQuery } from '@tanstack/react-query'

import { getAllHomeData } from '@/services/homeService'
import useLocaleStore from '@/store/locale'

export function useHome() {
  const { selectedLocale } = useLocaleStore()

  const {
    data: homeData,
    error,
    isLoading,
  } = useQuery({
    queryFn: async () => await getAllHomeData(selectedLocale),
    queryKey: ['homeData', selectedLocale],
    staleTime: 5 * 60 * 1000,
  })

  return {
    error,
    homeData,
    isLoading,
  }
}
