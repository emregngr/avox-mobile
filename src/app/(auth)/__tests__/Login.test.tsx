import { fireEvent, render, waitFor } from '@testing-library/react-native'
import { router, useLocalSearchParams } from 'expo-router'
import type { ReactNode } from 'react'
import React from 'react'
import { useForm } from 'react-hook-form'
import { SafeAreaProvider } from 'react-native-safe-area-context'

import Login from '@/app/(auth)/login'
import { useEmailLogin } from '@/hooks/services/useAuth'
import useLocaleStore from '@/store/locale'

jest.mock('@/locales/i18next', () => ({
  getLocale: (key: string) => key,
}))

jest.mock('react-hook-form')

const mockedUseForm = useForm as jest.MockedFunction<typeof useForm>

jest.mock('@/store/locale')

const mockedUseLocaleStore = useLocaleStore as jest.MockedFunction<typeof useLocaleStore>

jest.mock('@/components/common', () => {
  const { View, Text, TouchableOpacity, TextInput } = require('react-native')

  return {
    Header: ({ title, backIconOnPress }: { title: string; backIconOnPress: () => void }) => (
      <>
        <Text testID="header-title">{title}</Text>
        <TouchableOpacity onPress={backIconOnPress} testID="back-button">
          <Text>Back</Text>
        </TouchableOpacity>
      </>
    ),

    SafeLayout: ({ children }: { children: ReactNode }) => (
      <View testID="safe-layout">{children}</View>
    ),

    TextInputField: ({ label, editable, ...props }: { label: string; editable: boolean }) => (
      <View>
        <Text>{label}</Text>
        <TextInput {...props} accessibilityLabel={label} editable={editable} />
      </View>
    ),

    ThemedGradientButton: ({
      onPress,
      label,
      disabled,
      loading,
    }: {
      onPress: () => void
      label: string
      disabled: boolean
      loading: boolean
    }) => (
      <TouchableOpacity disabled={disabled || loading} onPress={onPress} testID="login-button">
        <Text>{loading ? 'Loading...' : label}</Text>
      </TouchableOpacity>
    ),

    ThemedButtonText: ({
      onPress,
      label,
      disabled,
      loading,
    }: {
      onPress: () => void
      label: string
      disabled: boolean
      loading: boolean
    }) => (
      <TouchableOpacity
        disabled={disabled || loading}
        onPress={onPress}
        testID={`button-${label}`}
      />
    ),

    ThemedText: ({ children, ...props }: { children: string }) => (
      <Text {...props}>{children}</Text>
    ),
  }
})

jest.mock('@/hooks/services/useAuth')

const mockedUseEmailLogin = useEmailLogin as jest.MockedFunction<typeof useEmailLogin>

const mockedUseLocalSearchParams = useLocalSearchParams as jest.MockedFunction<
  typeof useLocalSearchParams
>

const mockedReset = jest.fn()

const renderWithSafeAreaProvider = (component: ReactNode) =>
  render(
    <SafeAreaProvider
      initialMetrics={{
        insets: { top: 0, left: 0, right: 0, bottom: 0 },
        frame: { x: 0, y: 0, width: 0, height: 0 },
      }}
    >
      {component}
    </SafeAreaProvider>,
  )

beforeEach(() => {
  mockedUseLocaleStore.mockReturnValue({ selectedLocale: 'en' })

  mockedUseEmailLogin.mockReturnValue({
    isPending: false,
    mutateAsync: mockedUseEmailLogin,
  } as any)

  mockedUseLocalSearchParams.mockReturnValue({
    isRegisterParam: 'false',
  })

  mockedUseForm.mockReturnValue({
    control: {},
    handleSubmit: jest.fn(
      callback => () => callback({ email: 'test@example.com', password: 'password123' }),
    ),
    formState: { isValid: true },
    reset: mockedReset,
  } as any)
})

describe('Login Screen', () => {
  it('should render correctly', () => {
    const { getAllByText, getByLabelText, getByTestId } = renderWithSafeAreaProvider(<Login />)
    expect(getAllByText('login').length).toBeGreaterThan(0)
    expect(getByLabelText('email')).toBeTruthy()
    expect(getByLabelText('password')).toBeTruthy()
    expect(getByTestId('login-button')).not.toBeDisabled()
  })

  it('should call login mutation on submit', async () => {
    const handleSubmitMock = jest.fn(
      callback => () => callback({ email: 'test@example.com', password: 'password123' }),
    )
    mockedUseForm.mockReturnValue({
      control: {},
      handleSubmit: handleSubmitMock,
      formState: { isValid: true },
      reset: mockedReset,
    } as any)

    const { getByTestId } = renderWithSafeAreaProvider(<Login />)
    const loginButton = getByTestId('login-button')
    fireEvent.press(loginButton)

    await waitFor(() => {
      expect(handleSubmitMock).toHaveBeenCalled()
      expect(mockedUseEmailLogin).toHaveBeenCalledWith({
        email: 'test@example.com',
        password: 'password123',
      })
    })
  })

  it('should disable button when form is invalid after submission attempt', async () => {
    mockedUseForm.mockReturnValue({
      control: {},
      handleSubmit: jest.fn(callback => () => callback({})),
      formState: { isValid: false },
      reset: mockedReset,
    } as any)

    const { getByTestId } = renderWithSafeAreaProvider(<Login />)
    const loginButton = getByTestId('login-button')
    expect(loginButton).not.toBeDisabled()

    fireEvent.press(loginButton)

    await waitFor(() => {
      expect(loginButton).toBeDisabled()
    })
  })

  it('should show loading state and disable inputs', () => {
    ;(require('@/hooks/services/useAuth').useEmailLogin as jest.Mock).mockReturnValue({
      isPending: true,
      mutateAsync: mockedUseEmailLogin,
    })
    const { getByTestId, getByText, getByLabelText } = renderWithSafeAreaProvider(<Login />)

    expect(getByTestId('login-button')).toBeDisabled()
    expect(getByText('Loading...')).toBeTruthy()
    expect(getByLabelText('email').props.editable).toBe(false)
    expect(getByLabelText('password').props.editable).toBe(false)
  })

  it('should navigate to forgot-password', () => {
    const { getByTestId } = renderWithSafeAreaProvider(<Login />)
    const forgotPasswordButton = getByTestId('button-forgotPassword')
    fireEvent.press(forgotPasswordButton)
    expect(mockedReset).toHaveBeenCalled()
    expect(router.navigate).toHaveBeenCalledWith('/forgot-password')
  })

  it('should navigate to register', () => {
    const { getByTestId } = renderWithSafeAreaProvider(<Login />)
    const createAccountButton = getByTestId('button-createAccount')
    fireEvent.press(createAccountButton)
    expect(mockedReset).toHaveBeenCalled()
    expect(router.navigate).toHaveBeenCalledWith({
      params: { isLoginParam: 'true' },
      pathname: '/register',
    })
  })

  it('should navigate back to auth screen by default', () => {
    const { getByTestId } = renderWithSafeAreaProvider(<Login />)
    const backButton = getByTestId('back-button')
    fireEvent.press(backButton)
    expect(mockedReset).toHaveBeenCalled()
    expect(router.replace).toHaveBeenCalledWith('/auth')
  })

  it('should navigate back to register screen if coming from there', () => {
    ;(require('expo-router').useLocalSearchParams as jest.Mock).mockReturnValue({
      isRegisterParam: 'true',
    })
    const { getByTestId } = renderWithSafeAreaProvider(<Login />)
    const backButton = getByTestId('back-button')
    fireEvent.press(backButton)
    expect(mockedReset).toHaveBeenCalled()
    expect(router.navigate).toHaveBeenCalledWith('/register')
  })
})

describe('Login Screen Snapshot', () => {
  it('should render the Login successfully', () => {
    const { toJSON } = renderWithSafeAreaProvider(<Login />)

    expect(toJSON()).toMatchSnapshot()
  })
})
