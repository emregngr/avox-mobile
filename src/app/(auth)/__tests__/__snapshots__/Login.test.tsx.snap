// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Login Screen Snapshot should render the Login successfully 1`] = `
<View
  testID="safe-layout"
>
  <Text
    testID="header-title"
  >
    login
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="back-button"
  >
    <Text>
      Back
    </Text>
  </View>
  <RCTScrollView
    bottomOffset={50}
    contentContainerClassName="px-4"
    contentContainerStyle={
      {
        "paddingBottom": 20,
        "paddingTop": 64,
      }
    }
    keyboardDismissMode="on-drag"
    keyboardShouldPersistTaps="handled"
    showsVerticalScrollIndicator={false}
  >
    <View>
      <View
        className="items-center my-8"
      >
        <View
          cachePolicy="memory-disk"
          contentFit="contain"
          source={
            {
              "default": "mocked-test-file",
            }
          }
          style={
            {
              "borderRadius": 50,
              "height": 100,
              "width": 100,
            }
          }
          testID="mocked-image"
          transition={0}
        />
      </View>
      <View
        className="gap-y-2"
      >
        <View>
          <Text>
            email
          </Text>
          <TextInput
            accessibilityLabel="email"
            control={{}}
            editable={true}
            keyboardType="email-address"
            name="email"
            onSubmitEditing={[Function]}
            placeholder="emailPlaceholder"
            returnKeyType="next"
            secureTextEntry={false}
            showToggle={false}
            testID="email"
          />
        </View>
        <View>
          <Text>
            password
          </Text>
          <TextInput
            accessibilityLabel="password"
            control={{}}
            editable={true}
            keyboardType="default"
            name="password"
            onSubmitEditing={[Function]}
            placeholder="passwordPlaceholder"
            returnKeyType="done"
            secureTextEntry={true}
            showToggle={true}
            testID="password"
          />
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="button-forgotPassword"
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="login-button"
      >
        <Text>
          login
        </Text>
      </View>
      <View
        className="flex-row justify-center items-center mt-6"
      >
        <Text
          color="text-70"
          type="body2"
        >
          dontHaveAccount
        </Text>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="button-createAccount"
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
